cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityManager.java

```diff
@@ -0,0 +1,422 @@
+      LOG.error("Stopping pipelines failed for capability {} with exception {}", capability, ex);
+    }
+    //programs(services) will be stopped by SystemProgramManagementService
+    LOG.debug("Capability {} disabled.", capability);
+  }
+
+  private void deleteCapability(CapabilityConfig capabilityConfig) {
+    String capability = capabilityConfig.getCapability();
+    if (capabilityStatusMap.get(capability) == CapabilityStatus.ENABLED) {
+      LOG.error("Deleting capability {} failed. Capability should be disabled before deleting.", capability);
+      return;
+    }
+    //delete pipelines
+    try {
+      deletePipelines(capability);
+    } catch (Exception ex) {
+      LOG.error("Deleting pipelines failed for capability {} with exception {}", capability, ex);
+    }
+    //delete applications
+    for (SystemApplication application : capabilityConfig.getApplications()) {
+      ApplicationId applicationId = new ApplicationId(application.getNamespace(), application.getName(),
+                                                      application.getVersion());
+      try {
+        applicationLifecycleService.removeApplication(applicationId);
+      } catch (Exception exception) {
+        LOG.error("Deleting application {} failed with exception {}", applicationId, exception);
+      }
+    }
+    capabilityStatusMap.remove(capability);
+    LOG.debug("Capability {} deleted.", capability);
+  }
+
+  private void deletePipelines(String capability) throws Exception {
+    doForAllApps(capability, this::deleteAppWithRetry);
+  }
+
+  private void stopPipelines(String capability) throws Exception {
+    doForAllApps(capability, this::stopPrograms);
+  }
+
+  private void deployAllApps(String capability, List<SystemApplication> applications) {
+    if (applications.isEmpty()) {
+      LOG.debug("Capability {} do not have apps associated with it", capability);
+      return;
+    }
+    for (SystemApplication application : applications) {
+      doWithRetry(application, this::deployApp);
+    }
+  }
+
+  private void deployApp(SystemApplication application) {
+    String version = application.getVersion() == null ? ApplicationId.DEFAULT_VERSION : application.getVersion();
+    ApplicationId applicationId = new ApplicationId(application.getNamespace(), application.getName(), version);
+    LOG.debug("Deploying app {}", applicationId);
+    try {
+      if (isAppDeployed(applicationId)) {
+        //Already deployed.
+        LOG.debug("Application {} is already deployed", applicationId);
+        return;
+      }
+      String configString = application.getConfig() == null ? null : GSON.toJson(application.getConfig());
+      applicationLifecycleService
+        .deployApp(applicationId.getParent(), applicationId.getApplication(), applicationId.getVersion(),
+                   application.getArtifact(), configString, NOOP_PROGRAM_TERMINATOR, null, null);
+    } catch (Exception ex) {
+      throw new RetryableException(ex);
```

Comment:
Is everything exception retryable?

Reply:
We could restrict it to not retry for UnauthorizedException, InvalidArtifactException . Only retrying for ArtifactNotFoundException may not be sufficient I think. Not sure what the exception is in case system is not fully ready.

Reply:
Added some checks for exceptions.